File: DSA/DSA02003 DI CHUYỂN TRONG MÊ CUNG 1.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
bool found = false;
void go_right(vector<vector<int>> &A, int i, int j, string s);

void go_down(vector<vector<int>> &A, int i, int j, string s)
{
    if (i+j == n*2-2)
    {
        cout << s << " ";
        found = true;
        return;
    }

    if (i+1 == n || A[i+1][j] == 0)
    {
        return;
    }
    
    s += "D";
    i++;
    go_down(A, i, j, s);
    go_right(A, i, j, s);
}

void go_right(vector<vector<int>> &A, int i, int j, string s)
{
    if (j+1 == n || A[i][j+1] == 0)
    {
        return;
    }

    s += "R";
    j++;
    go_down(A, i, j, s);
    go_right(A, i, j, s);
}

void submit()
{
    cin >> n;
    vector<vector<int>> A(n, vector<int> (n));

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin >> A[i][j];
        }
    }

    if (A[0][0] != 0)
    {
        int i = 0, j = 0;
        string s;
        go_down(A, i, j, s);
        go_right(A, i, j, s);
        cout << endl;
    }

    if (!found)
    {
        cout << -1 << endl;
    }
    found = false;
}   

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();

    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}

File: DSA_python/DSA02003 DI CHUYỂN TRONG MÊ CUNG 1.py
def go_down(A, i, j, s):
    global found
    if i == n-1 and j == n-1:
        print(s, end=" ")
        found = True
        return

    if i+1 == n or A[i+1][j] == 0:
        return;

    s += "D"
    save = s
    i += 1
    go_down(A, i, j, s)
    go_right(A, i, j, s)

def go_right(A, i, j, s):
    if j+1 == n or A[i][j+1] == 0:
        return;

    s += "R"
    save = s
    j += 1
    go_down(A, i, j, s)
    go_right(A, i, j, s)

def submit():
    global n, found
    n = int(input())
    found = False
    A = []
    for i in range(n):
        A.append(list(map(int, input().split())))

    if A[0][0] != 0:
        i = j = 0
        s = ""
        go_down(A, i, j, s)
        go_right(A, i, j, s)
        print()
    
    if found == False:
        print(-1)
    found = False

test = int(input())
for i in range(test):
    submit()